src/Simulation/qdk_sim_rs/src/c_api.rs

```diff
@@ -0,0 +1,443 @@
+    as_capi_err(|| apply(sim_id, &[idx], |model| &model.s_adj))
+}
+
+/// Applies the `T` operation acting on a given qubit to a given simulator,
+/// using the currently set noise model.
+#[no_mangle]
+pub fn t(sim_id: usize, idx: usize) -> i64 {
+    as_capi_err(|| apply(sim_id, &[idx], |model| &model.t))
+}
+
+/// Applies the `Adjoint T` operation acting on a given qubit to a given simulator,
+/// using the currently set noise model.
+#[no_mangle]
+pub fn t_adj(sim_id: usize, idx: usize) -> i64 {
+    as_capi_err(|| apply(sim_id, &[idx], |model| &model.t_adj))
+}
+
+/// Applies the `CNOT` operation acting on two given qubits to a given simulator,
+/// using the currently set noise model.
+#[no_mangle]
+pub fn cnot(sim_id: usize, idx_control: usize, idx_target: usize) -> i64 {
+    as_capi_err(|| apply(sim_id, &[idx_control, idx_target], |model| &model.cnot))
+}
+
+/// Measures a single qubit in the $Z$-basis, returning the result by setting
+/// the value at a given pointer.
+///
+/// # Safety
+/// This function is marked as unsafe as it is the caller's responsibility to
+/// ensure that `result_out` is a valid pointer, and that the memory referenced
+/// by `result_out` can be safely set.
+#[no_mangle]
+pub unsafe extern "C" fn m(sim_id: usize, idx: usize, result_out: *mut usize) -> i64 {
+    as_capi_err(|| {
+        let state = &mut *STATE.lock().unwrap();
+        if let Some(sim_state) = state.get_mut(&sim_id) {
+            let instrument = &sim_state.noise_model.z_meas;
+            let (result, new_state) = instrument.sample(&[idx], &sim_state.register_state);
+            sim_state.register_state = new_state;
+            *result_out = result;
+            Ok(())
+        } else {
+            Err(format!("No simulator with id {} exists.", sim_id))
+        }
+    })
+}
+
+/// Gets the noise model corresponding to a particular name, serialized as a
+/// string representing a JSON object.
+///
+/// Currently recognized names:
+/// - `ideal`
+/// - `ideal_stabilizer`
+///
+/// # Safety
+/// As this is a C-language API, the Rust compiler cannot guarantee safety when
+/// calling into this function. The caller is responsible for ensuring that:
+///
+/// - `name` is a pointer to a null-terminated string of Unicode characters,
+///   encoded as UTF-8, and that the pointer remains valid for the lifetime of
+///   this call.
+/// - `noise_model_json` is a valid pointer whose lifetime extends for the
+///    duration of this function call.
+///
+/// After this call completes, this function guarantees that either of the two
+/// conditions below holds:
+///
+/// - The return value is negative, in which case calling `lasterr` will return
+///   an actionable error message, or
+/// - The return value is `0`, and `*noise_model_json` is a valid pointer to a
+///   null-terminated string of Unicode characters, encoded as UTF-8. In this
+///   case, the caller is considered to own the memory allocated for this
+///   string.
```

Comment:
> the caller is considered to own the memory allocated for this string

For the future, just in case, I would like to refer you to [Allocate and Deallocate In The Same Binary](https://github.com/microsoft/qsharp-runtime/blob/main/src/Qir/QIR-RT-API-Design-Guidelines.md#allocate-and-deallocate-in-the-same-binary) recommendation.
(something to keep in mind)

Reply:
Good point; in that case, I would expect that the caller can deallocate by passing their owned memory back to the Rust library, such that allocation and deallocation occur in the same binary. To make C# interoperability work, we would need to then call [`Marshal.PtrToStringUTF8`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.ptrtostringutf8) to copy out that string before calling the deallocator. I'll open an issue to track improving string handling across this API. Thanks!

Reply:
https://github.com/microsoft/qsharp-runtime/issues/732